src/Take.Elephant.Azure/AzureEventHubReceiverQueue.cs

```diff
@@ -1,148 +1,128 @@
+        {
+            return StartConsumerTaskIfNotAsync(cancellationToken);
         }
 
-        private async Task<bool> OpenIfRequiredAsync(CancellationToken cancellationToken)
+        private async Task StartConsumerTaskIfNotAsync(CancellationToken cancellationToken)
         {
-            if (_receiver != null) return false;
+            if (_consumerTask != null) return;
 
-            await _openSemaphore.WaitAsync(cancellationToken);
+            await _consumerStartSemaphore.WaitAsync(cancellationToken);
             try
             {
-                if (_receiver != null) return false;
-                _receiver = _eventHubClient.CreateReceiver(_consumerGroupName, _partitionId, _eventPosition, _receiverOptions);
-                return true;
+                if (_consumerTask == null)
+                {
+                    _consumerTask = Task
+                        .Factory
+                        .StartNew(
+                            () => ConsumeAsync(_cts.Token),
+                            TaskCreationOptions.LongRunning)
+                        .Unwrap();
+                }
             }
             finally
             {
-                _openSemaphore.Release();
+                _consumerStartSemaphore.Release();
             }
         }
 
-        private T CreateItem(EventData eventData)
-        {
-            if (eventData == null) return default;
-            return _serializer.Deserialize(Encoding.UTF8.GetString(eventData.Body.Array));            
-        }
-
-        internal class ExponentialInterval
+        private async Task ConsumeAsync(CancellationToken cancellationToken)
         {
-            private readonly TimeSpan _initialInterval;
-            private readonly TimeSpan _maxInterval;
-
-            public ExponentialInterval(TimeSpan initialInterval, TimeSpan maxInterval)
-            {
-                _initialInterval = initialInterval;
-                _maxInterval = maxInterval;
-            }
-            
-            public int Count { get; private set;  }
-            
-            public TimeSpan Interval
+            await foreach (var value in _consumer.ReadEventsAsync(cancellationToken))
             {
-                get
+                try
+                {
+                    var resultValue = _serializer.Deserialize(Encoding.UTF8.GetString(value.Data.Body.Span));
+                    await _channel.Writer.WriteAsync(resultValue, cancellationToken);
+                }
+                catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
+                {
+                    break;
+                }
+                catch (Exception ex)
                 {
-                    var timeout = TimeSpan.FromTicks(_initialInterval.Ticks * (int)Math.Pow(2, Count++));
-                    if (timeout < _maxInterval) return timeout;
-                    return _maxInterval;
+                    //todo: event handler
```

Comment:
We can't do that because this will cause the thread to stop consuming the Stream, I will make a event handler so we can act based on the exception.

Reply:
If possible, I would like to discuss this design choice before moving on with the PR. Currently, it appears that the only precedent for this design is in Kafka's impl. Discussing it now would be better as the more places we use it, the harder it will be to change it in the future.

I agree that it wouldn't be a great experience to have the consumer `new` up an instance of this class every time a socket dropped. However, I also don't agree that using events to signal errors is a good idea for the reasons I explained in the comment above. I feel like that this _could_ be one of those pragmatic choices where this solution works and we simply don't know a better one, but I'm unconvinced. 

One way to alleviate problem 1) is to call `ChannelWriter.Complete(Exception)` when `handler is null`, so that the consumer is forced to deal with any errors. This approach still doesn't solve problem 2) and also feels a bit invasive, forcing the consumer to subscribe to an event, especially since this would be 0% discoverable through code - instead we would have to rely on documentation. 

Ideally I would like for us to ditch events altogether. Earlier when I left that comment I had actually thought of an alternative way to use exceptions: the consumer thread could set a private variable `Exception _consumerException` or something like that. Methods that try to read from the channel (like `DequeueAsync`) would have to check this variable first and throw it when `!= null`. This is almost the same as the `ChannelWriter.Complete(Exception)` "solution" I proposed in the earlier comment (which is why I proposed it), except that it doesn't require tearing the thread down. I think that _could_ work. If the consumer wants, it is free to swallow to exception and continue execution. If it doesn't handle it though (maybe purposefully - maybe it's a library bug exception or maybe their retry logic decided that they've retried too many times and it's time to fail), then our `DisposeAsync` will eventually be called, which will eventually cause the consumer thread to be teared down.

Thoughts?

Reply:
Btw, is it possible to mark this as unresolved?